crate does not compile as no_std

```bash
cargo check --no-default-features --features alloc
```

Results in:

```bash
    Checking thingbuf v0.1.1 (/Users/chris/repos/github.com/hawkw/thingbuf)
error[E0433]: failed to resolve: could not find `unreachable_unchecked` in the crate root
   --> src/macros.rs:100:16
    |
100 |         crate::unreachable_unchecked!(@inner , format_args!(": {}", format_args!($($arg)*)))
    |                ^^^^^^^^^^^^^^^^^^^^^ could not find `unreachable_unchecked` in the crate root
    |
   ::: src/util/panic.rs:25:9
    |
25  |         unreachable_unchecked!("code compiled with no_std cannot unwind!")
    |         ------------------------------------------------------------------ in this macro invocation
    |
    = note: this error originates in the macro `unreachable_unchecked` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/lib.rs:169:9
    |
169 |         test_println!("Core::close");
    |         ---------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/lib.rs:185:9
    |
185 |         test_println!("push_ref");
    |         ------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/lib.rs:224:25
    |
224 |                         test_println!("claimed tail slot [{}]", idx);
    |                         -------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/lib.rs:235:33
    |
235 | ...                   test_println!("-> initialized");
    |                       ------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/lib.rs:240:33
    |
240 | ...                   test_println!("-> recycled");
    |                       ---------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/lib.rs:270:21
    |
270 |                     test_println!("channel full");
    |                     ----------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/lib.rs:285:9
    |
285 |         test_println!("pop_ref");
    |         ------------------------ in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/lib.rs:322:25
    |
322 |                         test_println!("claimed head slot [{}]", idx);
    |                         -------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/lib.rs:352:25
    |
352 |                         test_println!("--> channel full!");
    |                         ---------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/lib.rs:478:9
    |
478 |         test_println!("drop Ref<{}>", core::any::type_name::<T>());
    |         ---------------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
  --> src/macros.rs:7:17
   |
7  |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
   |                 ^^^^^^^
   |
  ::: src/wait.rs:79:9
   |
79 |         test_println!("WAKING TASK {:?} (from {:?})", self, thread::current());
   |         ---------------------------------------------------------------------- in this macro invocation
   |
   = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/wait/queue.rs:602:9
    |
602 |         test_println!("List::remove({:p})", node);
    |         ----------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/wait/queue.rs:578:9
    |
578 |         test_println!("List::dequeue({:?}) -> {:p}", new_state, last);
    |         ------------------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/wait/queue.rs:549:9
    |
549 |         test_println!("List::enqueue({:p})", waiter);
    |         -------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/wait/queue.rs:473:9
    |
473 |         test_println!("Waiter::remove({:p})", self);
    |         ------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/wait/queue.rs:444:9
    |
444 |         test_println!("WaitQueue::close()");
    |         ----------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/wait/queue.rs:399:9
    |
399 |         test_println!("WaitQueue::notify_slow(state: {})", state);
    |         --------------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/wait/queue.rs:371:9
    |
371 |         test_println!("WaitQueue::notify()");
    |         ------------------------------------ in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/wait/queue.rs:327:9
    |
327 |         test_println!("WaitQueue::continue_wait_slow({:p})", node);
    |         ---------------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/wait/queue.rs:302:9
    |
302 |         test_println!("WaitQueue::continue_wait({:p})", node);
    |         ----------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/wait/queue.rs:209:9
    |
209 |         test_println!("WaitQueue::start_wait_slow({:p})", node);
    |         ------------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/wait/queue.rs:189:9
    |
189 |         test_println!("WaitQueue::start_wait({:p})", node);
    |         -------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/wait/cell.rs:161:9
    |
161 |         test_println!("notifying; close={:?};", close);
    |         ---------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/wait/cell.rs:112:17
    |
112 |                 test_println!("-> was notified; state={:?}", actual);
    |                 ---------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
  --> src/macros.rs:7:17
   |
7  |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
   |                 ^^^^^^^
   |
  ::: src/wait/cell.rs:89:9
   |
89 |         test_println!("-> locked!");
   |         --------------------------- in this macro invocation
   |
   = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
  --> src/macros.rs:7:17
   |
7  |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
   |                 ^^^^^^^
   |
  ::: src/wait/cell.rs:66:9
   |
66 |         test_println!("registering waiter");
   |         ----------------------------------- in this macro invocation
   |
   = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
  --> src/macros.rs:7:17
   |
7  |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
   |                 ^^^^^^^
   |
  ::: src/util.rs:60:9
   |
60 |         test_println!("thread::yield_now()");
   |         ------------------------------------ in this macro invocation
   |
   = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
  --> src/macros.rs:7:17
   |
7  |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
   |                 ^^^^^^^
   |
  ::: src/util.rs:57:13
   |
57 |             test_println!("hint::spin_loop() (x{})", 1 << self.0);
   |             ----------------------------------------------------- in this macro invocation
   |
   = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
  --> src/macros.rs:7:17
   |
7  |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
   |                 ^^^^^^^
   |
  ::: src/util/mutex/spin_impl.rs:93:9
   |
93 |         test_println!("unlocked!");
   |         -------------------------- in this macro invocation
   |
   = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
  --> src/macros.rs:7:17
   |
7  |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
   |                 ^^^^^^^
   |
  ::: src/util/mutex/spin_impl.rs:48:9
   |
48 |         test_println!("-> locked {}!", core::any::type_name::<T>());
   |         ----------------------------------------------------------- in this macro invocation
   |
   = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
  --> src/macros.rs:7:17
   |
7  |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
   |                 ^^^^^^^
   |
  ::: src/util/mutex/spin_impl.rs:40:9
   |
40 |         test_println!("locking {}...", core::any::type_name::<T>());
   |         ----------------------------------------------------------- in this macro invocation
   |
   = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
    --> src/macros.rs:7:17
     |
7    |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
     |                 ^^^^^^^
     |
    ::: src/mpsc/async_impl.rs:1358:9
     |
1358 |         test_println!("SendRefFuture::drop({:p})", self);
     |         ------------------------------------------------ in this macro invocation
     |
     = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
    --> src/macros.rs:7:17
     |
7    |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
     |                 ^^^^^^^
     |
    ::: src/mpsc/async_impl.rs:1299:9
     |
1299 |         test_println!("SendRefFuture::poll({:p})", self);
     |         ------------------------------------------------ in this macro invocation
     |
     = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/mpsc.rs:484:9
    |
484 |         test_println!("notifying tx ({})", core::any::type_name::<N>());
    |         --------------------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/mpsc.rs:476:9
    |
476 |         test_println!("notifying rx ({})", core::any::type_name::<N>());
    |         --------------------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/mpsc.rs:395:21
    |
395 |                     test_println!("-> yield");
    |                     ------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/mpsc.rs:382:13
    |
382 |             test_println!("poll_recv_ref => loop");
    |             -------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/mpsc.rs:380:9
    |
380 |         test_println!("poll_recv_ref");
    |         ------------------------------ in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error: cannot find macro `println` in this scope
   --> src/macros.rs:7:17
    |
7   |                 println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
    |                 ^^^^^^^
    |
   ::: src/mpsc.rs:322:13
    |
322 |             test_println!("draining_queue");
    |             ------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/mpsc.rs:322:13
    |
322 |             test_println!("draining_queue");
    |             ------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/mpsc.rs:380:9
    |
380 |         test_println!("poll_recv_ref");
    |         ------------------------------ in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/mpsc.rs:382:13
    |
382 |             test_println!("poll_recv_ref => loop");
    |             -------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/mpsc.rs:395:21
    |
395 |                     test_println!("-> yield");
    |                     ------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/mpsc.rs:476:9
    |
476 |         test_println!("notifying rx ({})", core::any::type_name::<N>());
    |         --------------------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/mpsc.rs:484:9
    |
484 |         test_println!("notifying tx ({})", core::any::type_name::<N>());
    |         --------------------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
    --> src/macros.rs:11:34
     |
11   |                     crate::loom::thread::current().id(),
     |                                  ^^^^^^ could not find `thread` in `loom`
     |
    ::: src/mpsc/async_impl.rs:1299:9
     |
1299 |         test_println!("SendRefFuture::poll({:p})", self);
     |         ------------------------------------------------ in this macro invocation
     |
     = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
    --> src/macros.rs:11:34
     |
11   |                     crate::loom::thread::current().id(),
     |                                  ^^^^^^ could not find `thread` in `loom`
     |
    ::: src/mpsc/async_impl.rs:1358:9
     |
1358 |         test_println!("SendRefFuture::drop({:p})", self);
     |         ------------------------------------------------ in this macro invocation
     |
     = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
  --> src/macros.rs:11:34
   |
11 |                     crate::loom::thread::current().id(),
   |                                  ^^^^^^ could not find `thread` in `loom`
   |
  ::: src/util/mutex/spin_impl.rs:40:9
   |
40 |         test_println!("locking {}...", core::any::type_name::<T>());
   |         ----------------------------------------------------------- in this macro invocation
   |
   = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
  --> src/macros.rs:11:34
   |
11 |                     crate::loom::thread::current().id(),
   |                                  ^^^^^^ could not find `thread` in `loom`
   |
  ::: src/util/mutex/spin_impl.rs:48:9
   |
48 |         test_println!("-> locked {}!", core::any::type_name::<T>());
   |         ----------------------------------------------------------- in this macro invocation
   |
   = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
  --> src/macros.rs:11:34
   |
11 |                     crate::loom::thread::current().id(),
   |                                  ^^^^^^ could not find `thread` in `loom`
   |
  ::: src/util/mutex/spin_impl.rs:93:9
   |
93 |         test_println!("unlocked!");
   |         -------------------------- in this macro invocation
   |
   = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
  --> src/macros.rs:11:34
   |
11 |                     crate::loom::thread::current().id(),
   |                                  ^^^^^^ could not find `thread` in `loom`
   |
  ::: src/util.rs:57:13
   |
57 |             test_println!("hint::spin_loop() (x{})", 1 << self.0);
   |             ----------------------------------------------------- in this macro invocation
   |
   = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
  --> src/macros.rs:11:34
   |
11 |                     crate::loom::thread::current().id(),
   |                                  ^^^^^^ could not find `thread` in `loom`
   |
  ::: src/util.rs:60:9
   |
60 |         test_println!("thread::yield_now()");
   |         ------------------------------------ in this macro invocation
   |
   = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
  --> src/macros.rs:11:34
   |
11 |                     crate::loom::thread::current().id(),
   |                                  ^^^^^^ could not find `thread` in `loom`
   |
  ::: src/wait/cell.rs:66:9
   |
66 |         test_println!("registering waiter");
   |         ----------------------------------- in this macro invocation
   |
   = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
  --> src/macros.rs:11:34
   |
11 |                     crate::loom::thread::current().id(),
   |                                  ^^^^^^ could not find `thread` in `loom`
   |
  ::: src/wait/cell.rs:89:9
   |
89 |         test_println!("-> locked!");
   |         --------------------------- in this macro invocation
   |
   = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/wait/cell.rs:112:17
    |
112 |                 test_println!("-> was notified; state={:?}", actual);
    |                 ---------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/wait/cell.rs:161:9
    |
161 |         test_println!("notifying; close={:?};", close);
    |         ---------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/wait/queue.rs:189:9
    |
189 |         test_println!("WaitQueue::start_wait({:p})", node);
    |         -------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/wait/queue.rs:209:9
    |
209 |         test_println!("WaitQueue::start_wait_slow({:p})", node);
    |         ------------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/wait/queue.rs:302:9
    |
302 |         test_println!("WaitQueue::continue_wait({:p})", node);
    |         ----------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/wait/queue.rs:327:9
    |
327 |         test_println!("WaitQueue::continue_wait_slow({:p})", node);
    |         ---------------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/wait/queue.rs:371:9
    |
371 |         test_println!("WaitQueue::notify()");
    |         ------------------------------------ in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/wait/queue.rs:399:9
    |
399 |         test_println!("WaitQueue::notify_slow(state: {})", state);
    |         --------------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/wait/queue.rs:444:9
    |
444 |         test_println!("WaitQueue::close()");
    |         ----------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/wait/queue.rs:473:9
    |
473 |         test_println!("Waiter::remove({:p})", self);
    |         ------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/wait/queue.rs:549:9
    |
549 |         test_println!("List::enqueue({:p})", waiter);
    |         -------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/wait/queue.rs:578:9
    |
578 |         test_println!("List::dequeue({:?}) -> {:p}", new_state, last);
    |         ------------------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/wait/queue.rs:602:9
    |
602 |         test_println!("List::remove({:p})", node);
    |         ----------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
  --> src/macros.rs:11:34
   |
11 |                     crate::loom::thread::current().id(),
   |                                  ^^^^^^ could not find `thread` in `loom`
   |
  ::: src/wait.rs:79:9
   |
79 |         test_println!("WAKING TASK {:?} (from {:?})", self, thread::current());
   |         ---------------------------------------------------------------------- in this macro invocation
   |
   = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: use of undeclared crate or module `thread`
  --> src/wait.rs:79:61
   |
79 |         test_println!("WAKING TASK {:?} (from {:?})", self, thread::current());
   |                                                             ^^^^^^ use of undeclared crate or module `thread`

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/lib.rs:169:9
    |
169 |         test_println!("Core::close");
    |         ---------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: use of undeclared crate or module `std`
   --> src/lib.rs:170:12
    |
170 |         if std::thread::panicking() {
    |            ^^^ use of undeclared crate or module `std`

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/lib.rs:185:9
    |
185 |         test_println!("push_ref");
    |         ------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/lib.rs:224:25
    |
224 |                         test_println!("claimed tail slot [{}]", idx);
    |                         -------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/lib.rs:235:33
    |
235 | ...                   test_println!("-> initialized");
    |                       ------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/lib.rs:240:33
    |
240 | ...                   test_println!("-> recycled");
    |                       ---------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/lib.rs:270:21
    |
270 |                     test_println!("channel full");
    |                     ----------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/lib.rs:285:9
    |
285 |         test_println!("pop_ref");
    |         ------------------------ in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/lib.rs:322:25
    |
322 |                         test_println!("claimed head slot [{}]", idx);
    |                         -------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/lib.rs:352:25
    |
352 |                         test_println!("--> channel full!");
    |                         ---------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `thread` in `loom`
   --> src/macros.rs:11:34
    |
11  |                     crate::loom::thread::current().id(),
    |                                  ^^^^^^ could not find `thread` in `loom`
    |
   ::: src/lib.rs:478:9
    |
478 |         test_println!("drop Ref<{}>", core::any::type_name::<T>());
    |         ---------------------------------------------------------- in this macro invocation
    |
    = note: this error originates in the macro `test_println` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0412]: cannot find type `Box` in this scope
  --> src/mpsc/async_impl.rs:72:16
   |
72 |         slots: Box<[Slot<T>]>,
   |                ^^^ not found in this scope
   |
help: consider importing this struct
   |
1  | use alloc::boxed::Box;
   |

error[E0412]: cannot find type `Box` in this scope
   --> src/lib.rs:519:56
    |
519 |     pub(crate) fn make_boxed_array(capacity: usize) -> Box<[Self]> {
    |                                                        ^^^ not found in this scope
    |
help: consider importing this struct
    |
5   | use alloc::boxed::Box;
    |

error[E0392]: parameter `T` is never used
  --> src/mpsc/async_impl.rs:57:25
   |
57 |     pub struct Receiver<T, R = recycling::DefaultRecycle> {
   |                         ^ unused parameter
   |
   = help: consider removing `T`, referring to it in a field, or using a marker such as `PhantomData`
   = help: if you intended `T` to be a const parameter, use `const T: usize` instead

error[E0392]: parameter `T` is never used
  --> src/mpsc/async_impl.rs:66:23
   |
66 |     pub struct Sender<T, R = recycling::DefaultRecycle> {
   |                       ^ unused parameter
   |
   = help: consider removing `T`, referring to it in a field, or using a marker such as `PhantomData`
   = help: if you intended `T` to be a const parameter, use `const T: usize` instead

Some errors have detailed explanations: E0392, E0412, E0433.
For more information about an error, try `rustc --explain E0392`.
error: could not compile `thingbuf` due to 85 previous errors
```

At a glance:

* There are `println`s gated behind `if cfg!(test) || cfg!(all(thingbuf_trace, feature = "std"))` instead of `#[cfg(any(test, all(thingbuf_trace, feature = "std")))]`.
* `Core::close` references `std::thread::panicking()`
* `Box` is used without being imported from `alloc::boxed::Box`.

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -33,9 +33,9 @@ jobs:
         override: true
         components: rustfmt
       uses: actions-rs/cargo@v1
-        with:
-          command: check
-          args: --no-default-features --features alloc
+      with:
+        command: check
+        args: --no-default-features --features alloc
 
   tests:
     name: Tests

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -22,7 +22,7 @@ name: Tests
 jobs:
   build_no_std:
     name: cargo check (no-std)
-    runs_on: ubuntu-latest
+    runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v2
     - name: Install toolchain

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -20,6 +20,22 @@ env:
 
 name: Tests
 jobs:
+  build_no_std:
+    name: cargo check (no-std)
+    runs_on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - name: Install toolchain
+      uses: actions-rs/toolchain@v1
+      with:
+        profile: minimal
+        toolchain: stable
+        override: true
+        components: rustfmt
+      uses: actions-rs/cargo@v1
+        with:
+          command: check
+          args: --no-default-features --features alloc
 
   tests:
     name: Tests

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -21,7 +21,10 @@ env:
 name: Tests
 jobs:
   build_no_std:
-    name: cargo check (no-std)
+    name: Check no_std
+    strategy:
+      matrix:
+        feature: [alloc, static]
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v2
@@ -35,7 +38,7 @@ jobs:
     - uses: actions-rs/cargo@v1
       with:
         command: check
-        args: --no-default-features --features alloc
+        args: --no-default-features --features ${{ matrix.feature }}
 
   tests:
     name: Tests

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -32,7 +32,7 @@ jobs:
         toolchain: stable
         override: true
         components: rustfmt
-      uses: actions-rs/cargo@v1
+    - uses: actions-rs/cargo@v1
       with:
         command: check
         args: --no-default-features --features alloc

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -167,7 +167,7 @@ impl Core {
 
     fn close(&self) -> bool {
         test_println!("Core::close");
-        if std::thread::panicking() {
+        if crate::util::panic::panicking() {
             return false;
         }
         test_dbg!(self.tail.fetch_or(self.closed, SeqCst) & self.closed == 0)

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -516,7 +516,7 @@ unsafe impl<T: Send> Sync for Ref<'_, T> {}
 
 impl<T> Slot<T> {
     #[cfg(feature = "alloc")]
-    pub(crate) fn make_boxed_array(capacity: usize) -> Box<[Self]> {
+    pub(crate) fn make_boxed_array(capacity: usize) -> alloc::boxed::Box<[Self]> {
         (0..capacity).map(|i| Slot::new(i)).collect()
     }
 

--- a/src/macros.rs
+++ b/src/macros.rs
@@ -1,19 +1,18 @@
 macro_rules! test_println {
     ($($arg:tt)*) => {
-        if cfg!(test) || cfg!(all(thingbuf_trace, feature = "std")) {
-            if crate::util::panic::panicking() {
-                // getting the thread ID while panicking doesn't seem to play super nicely with loom's
-                // mock lazy_static...
-                println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
-            } else {
-                crate::loom::traceln(format_args!(
-                    "[{:?} {:>30}:{:<3}] {}",
-                    crate::loom::thread::current().id(),
-                    file!(),
-                    line!(),
-                    format_args!($($arg)*),
-                ));
-            }
+        #[cfg(all(feature = "std", any(thingbuf_trace, test)))]
+        if crate::util::panic::panicking() {
+            // getting the thread ID while panicking doesn't seem to play super nicely with loom's
+            // mock lazy_static...
+            println!("[PANIC {:>30}:{:<3}] {}", file!(), line!(), format_args!($($arg)*))
+        } else {
+            crate::loom::traceln(format_args!(
+                "[{:?} {:>30}:{:<3}] {}",
+                crate::loom::thread::current().id(),
+                file!(),
+                line!(),
+                format_args!($($arg)*),
+            ));
         }
     }
 }
@@ -22,7 +21,7 @@ macro_rules! test_dbg {
     ($e:expr) => {
         match $e {
             e => {
-                #[cfg(any(test, all(thingbuf_trace, feature = "std")))]
+                #[cfg(all(feature = "std", any(thingbuf_trace, test)))]
                 test_println!("{} = {:?}", stringify!($e), &e);
                 e
             }
@@ -40,7 +39,7 @@ macro_rules! assert_dbg {
     };
     (@$e:expr, $($msg:tt)+) => {
         {
-            #[cfg(any(test, all(thingbuf_trace, feature = "std")))]
+            #[cfg(all(feature = "std", any(thingbuf_trace, test)))]
             test_println!("ASSERT: {}{}", stringify!($e), format_args!($($msg)*));
             assert!($e, $($msg)*);
             test_println!("-> ok");
@@ -58,7 +57,7 @@ macro_rules! assert_eq_dbg {
     };
     (@ $a:expr, $b:expr, $($msg:tt)+) => {
         {
-            #[cfg(any(test, all(thingbuf_trace, feature = "std")))]
+            #[cfg(all(feature = "std", any(thingbuf_trace, test)))]
             test_println!("ASSERT: {} == {}{}", stringify!($a), stringify!($b), format_args!($($msg)*));
             assert_eq!($a, $b, $($msg)*);
             test_println!("-> ok");

--- a/src/macros.rs
+++ b/src/macros.rs
@@ -95,29 +95,32 @@ macro_rules! fmt_bits {
 
 #[allow(unused_macros)]
 macro_rules! unreachable_unchecked {
+    (@inner $msg:expr) => {
+        {
+            #[cfg(debug_assertions)] {
+                panic!(
+                    "internal error: entered unreachable code{}\n\n\
+                    /!\\ EXTREMELY SERIOUS WARNING /!\\\n
+                    This code should NEVER be entered; in release mode, this would \
+                    have been an `unreachable_unchecked` hint. The fact that this \
+                    occurred means something VERY bad is going on. \n\
+                    Please contact the `thingbuf` maintainers immediately. Sorry!",
+                    $msg,
+                );
+            }
+            #[cfg(not(debug_assertions))]
+            unsafe {
+                core::hint::unreachable_unchecked();
+            }
+
+        }
+    };
+
     ($($arg:tt)+) => {
-        crate::unreachable_unchecked!(@inner , format_args!(": {}", format_args!($($arg)*)))
+        unreachable_unchecked!(@inner format_args!(": {}", format_args!($($arg)*)))
     };
 
     () => {
-        crate::unreachable_unchecked!(@inner ".")
-    };
-
-    (@inner $msg:expr) => {
-        #[cfg(debug_assertions)] {
-            panic!(
-                "internal error: entered unreachable code{}\n\n\
-                /!\\ EXTREMELY SERIOUS WARNING /!\\\n
-                This code should NEVER be entered; in release mode, this would \
-                have been an `unreachable_unchecked` hint. The fact that this \
-                occurred means something VERY bad is going on. \n\
-                Please contact the `thingbuf` maintainers immediately. Sorry!",
-                $msg,
-            );
-        }
-        #[cfg(not(debug_assertions))]
-        unsafe {
-            core::hint::unreachable_unchecked();
-        }
+        unreachable_unchecked!(@inner ".")
     };
 }

--- a/src/mpsc/async_impl.rs
+++ b/src/mpsc/async_impl.rs
@@ -17,6 +17,7 @@ feature! {
     #![feature = "alloc"]
 
     use crate::loom::sync::Arc;
+    use alloc::boxed::Box;
 
     /// Returns a new asynchronous multi-producer, single consumer (MPSC)
     /// channel with the provided capacity.

--- a/src/util/mutex/spin_impl.rs
+++ b/src/util/mutex/spin_impl.rs
@@ -28,6 +28,7 @@ pub(crate) const fn const_mutex<T>(data: T) -> Mutex<T> {
 }
 
 impl<T> Mutex<T> {
+    #[cfg(loom)]
     pub(crate) fn new(data: T) -> Self {
         Self {
             locked: AtomicBool::new(false),

--- a/src/util/panic.rs
+++ b/src/util/panic.rs
@@ -21,7 +21,7 @@ mod inner {
         Ok(f())
     }
 
-    pub(crate) fn resume_unwind(payload: ()) -> ! {
+    pub(crate) fn resume_unwind(_: ()) -> ! {
         unreachable_unchecked!("code compiled with no_std cannot unwind!")
     }
 }

